### PR TITLE
Fix sctp compile errors

### LINKED ISSUE #19371: Master branch build is broken with enable-sctp option
I built it this way:
```
./Configure --prefix=$HOME/tmp/openssl --openssldir=/etc/pki/tls \
    zlib enable-rfc3779 enable-sctp enable-cms no-mdc2 no-ec2m no-sm2 no-sm4 \
    enable-buildtest-c++ shared '-DDEVRANDOM="\"/dev/urandom\""'
make all
```
and the build breaks this way:
```
[..]
gcc  -I. -Iinclude  -DAES_ASM -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSLDIR="\"/etc/pki/tls\"" -DENGINESDIR="\"/home/tmp/openssl/lib64/engines-3\"" -DMODULESDIR="\"/home/tmp/openssl/lib64/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DZLIB -DNDEBUG -DDEVRANDOM="\"/dev/urandom\"" -MMD -MF ssl/record/libssl-lib-rec_layer_d1.d.tmp -MT ssl/record/libssl-lib-rec_layer_d1.o -c -o ssl/record/libssl-lib-rec_layer_d1.o ssl/record/rec_layer_d1.c
ssl/record/rec_layer_d1.c: In function 'dtls_buffer_record':
ssl/record/rec_layer_d1.c:140:40: error: 'ssl' undeclared (first use in this function)
  140 |     if (BIO_dgram_is_sctp(SSL_get_rbio(ssl)) &&
      |                                        ^~~
ssl/record/rec_layer_d1.c:140:40: note: each undeclared identifier is reported only once for each function it appears in
ssl/record/rec_layer_d1.c: In function 'dtls_unbuffer_record':
ssl/record/rec_layer_d1.c:178:44: warning: passing argument 1 of 'SSL_get_rbio' from incompatible pointer type [-Wincompatible-pointer-types]
  178 |         if (BIO_dgram_is_sctp(SSL_get_rbio(s))) {
      |                                            ^
      |                                            |
      |                                            SSL_CONNECTION * {aka struct ssl_connection_st *}
In file included from ssl/record/../ssl_local.h:26,
                 from ssl/record/rec_layer_d1.c:12:
include/openssl/ssl.h:1655:37: note: expected 'const SSL *' {aka 'const struct ssl_st *'} but argument is of type 'SSL_CONNECTION *' {aka 'struct ssl_connection_st *'}
 1655 | __owur BIO *SSL_get_rbio(const SSL *s);
      |                          ~~~~~~~~~~~^
ssl/record/rec_layer_d1.c:179:35: warning: passing argument 1 of 'SSL_get_rbio' from incompatible pointer type [-Wincompatible-pointer-types]
  179 |             BIO_ctrl(SSL_get_rbio(s), BIO_CTRL_DGRAM_SCTP_SET_RCVINFO,
      |                                   ^
      |                                   |
      |                                   SSL_CONNECTION * {aka struct ssl_connection_st *}
include/openssl/ssl.h:1655:37: note: expected 'const SSL *' {aka 'const struct ssl_st *'} but argument is of type 'SSL_CONNECTION *' {aka 'struct ssl_connection_st *'}
 1655 | __owur BIO *SSL_get_rbio(const SSL *s);
      |                          ~~~~~~~~~~~^
make[1]: *** [Makefile:20315: ssl/record/libssl-lib-rec_layer_d1.o] Error 1
make[1]: Leaving directory '/home/devel/git/openssl'
make: *** [Makefile:3392: build_sw] Error 2
```

Like cause is commit eddb067e2ce82bc2ea104b3ab5286fe334c0525d

### PR DESCRIPTION
**Pull request by slontis**
Fixes #19371

running config with 'enable-sctp' gave compiler errors.

### CODE REVIEW

**Comment by slontis** [comment]
Tests passed locally with these changes..

To add this to our ci loop it would need to do the following:

sudo apt install libsctp-dev
sysctl -w net.sctp.auth_enable=1

**Review by t8m — DISMISSED** [review]

**Review by paulidale — CHANGES_REQUESTED** [review]
The fix here looks good but I would expect a `enable-sctp` build to be added to the CIs.  Possibly in parallel with or replacing the runchecker `no-sctp` build.

`sudo` works in CIs. It might need a conditional build step so this only gets added for the `enable-sctp` build.
If added to one of the non-matrix builds, the install/setup sets would be easy to add.

**Comment by slontis** [comment]
Did you see the preconditions to doing this pauli?

File: ssl/record/rec_layer_d1.c [hunk 1]
```diff
@@ -136,12 +136,16 @@ int dtls_buffer_record(SSL_CONNECTION *s, TLS_RECORD *rec)
...
-    if (BIO_dgram_is_sctp(SSL_get_rbio(ssl)) &&
-        (SSL_get_state(ssl) == TLS_ST_SR_FINISHED
-         || SSL_get_state(ssl) == TLS_ST_CR_FINISHED)) {
-        BIO_ctrl(SSL_get_rbio(ssl), BIO_CTRL_DGRAM_SCTP_GET_RCVINFO,
-                 sizeof(rdata->recordinfo), &rdata->recordinfo);
+    {
+        SSL *ssl = SSL_CONNECTION_GET_SSL(s);
+
+        /* Store bio_dgram_sctp_rcvinfo struct */
+        if (BIO_dgram_is_sctp(SSL_get_rbio(ssl)) &&
```
[the remark this answers is not in the document]
Comment (slontis) [reply ssl/record/rec_layer_d1.c:143]: ok, changed the other file to use s->wbio also.

Comment (mattcaswell) [inline ssl/record/rec_layer_d1.c:143]: `s->rbio`

File: ssl/record/rec_layer_d1.c [hunk 2]
```diff
@@ -136,12 +136,16 @@ int dtls_buffer_record(SSL_CONNECTION *s, TLS_RECORD *rec)
...
+            (SSL_get_state(ssl) == TLS_ST_SR_FINISHED
+             || SSL_get_state(ssl) == TLS_ST_CR_FINISHED)) {
+            BIO_ctrl(SSL_get_rbio(ssl), BIO_CTRL_DGRAM_SCTP_GET_RCVINFO,
```
Comment (mattcaswell) [inline ssl/record/rec_layer_d1.c:146]: `s->rbio`

File: ssl/record/rec_layer_d1.c [hunk 3]
```diff
@@ -136,12 +136,16 @@ int dtls_buffer_record(SSL_CONNECTION *s, TLS_RECORD *rec)
...
-         || SSL_get_state(ssl) == TLS_ST_CR_FINISHED)) {
-        BIO_ctrl(SSL_get_rbio(ssl), BIO_CTRL_DGRAM_SCTP_GET_RCVINFO,
-                 sizeof(rdata->recordinfo), &rdata->recordinfo);
+    {
+        SSL *ssl = SSL_CONNECTION_GET_SSL(s);
+
+        /* Store bio_dgram_sctp_rcvinfo struct */
+        if (BIO_dgram_is_sctp(SSL_get_rbio(ssl)) &&
+            (SSL_get_state(ssl) == TLS_ST_SR_FINISHED
+             || SSL_get_state(ssl) == TLS_ST_CR_FINISHED)) {
```
Comment (mattcaswell) [inline ssl/record/rec_layer_d1.c:145]: It feels wrong to me to have to get the SSL object, in order call SSL_get_state() which converts it back internally into an `SSL_CONNECTION`. Possibly we need an internal `ossl_statem_get_state()` function to return this. Or maybe I'm over thinking it. Not sure.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/record/rec_layer_d1.c (hunk 3 above)
Comment (t8m) [reply ssl/record/rec_layer_d1.c:145]: The conversion to SSL object should be basically free and the conversion to SSL_CONNECTION is just a NULL check so it should be fairly cheap. But yeah, we could have ossl_statem_get_state().

**Comment by bernd-edlinger** [comment]
Isn't the commit message mis-typed?
"sctp" instead of "stcp"

File: ssl/record/rec_layer_d1.c (hunk 3 above)
Comment (slontis) [reply ssl/record/rec_layer_d1.c]: Added function

**Comment by slontis** [comment]
Testing with a dummy action show that 'lksctp-tools' needed to be installed also..
The tests passed with this setup. The tests also failed if net.sctp.auth_enable was not set (indicating that the sctp is being used in the tests).

File: .github/workflows/run-checker-daily.yml [hunk 4]
```diff
@@ -137,6 +137,12 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
+    - name: Install Dependencies for sctp option
+      run:  |
+        sudo apt-get update
+        sudo apt-get -yq install lksctp-tools libsctp-dev
+        sudo sysctl -w net.sctp.auth_enable=1
+      if: matrix.opt == 'enable-sctp'
```
Comment (paulidale) [inline .github/workflows/run-checker-daily.yml:145]: I suggest putting this immediate under the `name:` line.

**Review by paulidale — APPROVED** [review]
With the `if` line moved up.

**Comment by paulidale** [comment]
@t8m reapprove?

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by hlandau** [comment]
Merged to master. Thank you.

**Pull request closed without merging**